docs/enhancements/image-service-cloud-authentication.md

```diff
@@ -0,0 +1,164 @@
+---
+title: image-service-cloud-authentication
+authors:
+  - "@carbonin"
+creation-date: 2021-10-07
+last-updated: 2021-10-07
+---
+
+# Image Service Cloud Authentication
+
+## Summary
+
+Authentication between the image service and assisted service when deployed
+in the cloud requires a new strategy because the existing tokens used in the cloud
+are too large and too short lived to be included in a pre-signed URL. Instead
+of the existing authentication used in the cloud, a password and expiration will be
+generated and recorded in the infraEnv when it is updated and on demand through a new
+endpoint. This password will be used as authentication for requests to assisted service
+specifically for the endpoints required by the image service.
+
+## Motivation
+
+When the assisted service is deployed in the cloud, it relies on Red Hat
+SSO for user authentication and authentication tokens from Red Hat SSO are not
+usable with pre-signed URLs for two main reasons:
+
+* Size - these tokens are typically ~2000 characters long
+* Expiration - these tokens have a maximum lifespan of 15 minutes
+
+### Goals
+
+* Define an authentication strategy suitable for pre-signed URLs in the cloud. This includes a
+token with a longer expiration (several hours) and a reasonable size (10s to 100s of characters)
+which will be included directly in a download URL
+
+### Non-Goals
+
+* Tie an image download URL to a Red Hat account or organization
+* Maintain public/private key pairs to generate and issue tokens
+* Add additional external identity management components
+
+## Proposal
+
+The new authentication mechanism will be a simple secret and expiration date
+both of which are tied to an infraEnv. This secret will be included as a
+parameter in the image download URL and assisted service authentication will
+pass if the secret in the URL matches the one currently in the infraEnv record
+and the request is processed before the expiration time. A new secret will be
```

Comment:
Instead of 1 hour or 4 hours, why not 24 hours, or a week?

Reply:
So even today if I get the URL and perhaps even start the download, that will fail if it's not done by the 4h mark? There's no graceful rotation?

Reply:
> What would the value be for including it unsigned? Would we expect someone to use it with SSO auth?

No, the idea was just to avoid duplicating information while also separating concerns. If we want a separate endpoint that is specifically meant to fetch the signed url and possibly (see comment below) refresh the token I wouldn't also want a signed url in the infraenv resource where we'd have to decide whether to re-implement the token refresh logic or knowingly give the user an expired or soon-expiring token.

Specifically I don't think it's a good idea to trigger a token refresh when we `GET` an infraenv because that could leave a user with an invalid URL that they reasonably thought would be valid for some time just because someone else queried the infraenv.

So I guess the options are give the user a URL that might no longer be valid when they query an infraenv (which is probably what we're doing today), or leave the URL without auth.

> It seems like a reasonable goal to not invalidate a URL that was generated 30s ago just because a new URL was retrieved.

So it sounds like we want something like an initial expiration and some lower time where we'd force a refresh if someone requests the url even if the token isn't actually expired yet. Say, if there's < 30 minutes left on a 4 hour token. Maybe even a parameter to force a refresh at any time.

Reply:
> So I guess the options are give the user a URL that might no longer be valid when they query an infraenv (which is probably what we're doing today), or leave the URL without auth.

The third option is to not put any download URL in the infraenv resource representation, and only make such URLs available from a dedicated endpoint. Then when someone uses that endpoint, they will get a URL that works for some documented period of time, and it won't affect previously-retrieved download URLs.